File: CPP/Backjoon/2263.cpp
#include <iostream>
#include <vector>

std::vector<int> infixes, postfixes;

void DFS(int infixStart, int infixEnd, int postfixHeadIndex) {
    if (infixStart > infixEnd) return;

    auto headValue = postfixes[postfixHeadIndex];
    // infix head index
    int infixHeadIndex = -1;
    for (int i = infixStart; i <= infixEnd; i++) {
        if (infixes[i] == headValue) {
            infixHeadIndex = i;
            break;
        }
    }

    int leftInfixStart = infixStart;
    int leftInfixEnd = infixHeadIndex - 1;

    int rightInfixStart = infixHeadIndex + 1;
    int rightInfixEnd = infixEnd;

    int left_postfixHeadIndex = postfixHeadIndex - (rightInfixEnd - rightInfixStart + 1) - 1;
    int right_postfixHeadIndex = postfixHeadIndex - 1;

    std::cout << headValue << " ";

    DFS(leftInfixStart, leftInfixEnd, left_postfixHeadIndex);
    DFS(rightInfixStart, rightInfixEnd, right_postfixHeadIndex);

}

void solve() {
    DFS(0, infixes.size() - 1, infixes.size() - 1);
}

int main() {
    int N;
    std::cin >> N;

    infixes = std::vector<int>(N, 0);
    postfixes = std::vector<int>(N, 0);

    for (int i = 0; i < N; i++) {
        int val; std::cin >> val;
        infixes[i] = val;
    }

    for (int i = 0; i < N; i++) {
        int val; std::cin >> val;
        postfixes[i] = val;
    }

    solve();

    return 0;
}

File: Python/backjoon/2263.py
import sys

sys.setrecursionlimit(10**5 + 1000)

from typing import List

N = int(input())

infixes = [*map(int, input().split())]
postfixes = [*map(int, input().split())]

def find(li: List[int], start: int, end: int, value: int):
    for index in range(start, end + 1):
        if li[index] == value:
            return index

    return -1 # ?

# postfixIndex는 현재 head가 어딘지 가리킴
def solve(infixStart: int, infixEnd: int, postfixHeadIndex: int):
    if infixStart > infixEnd:
        return

    head = postfixes[postfixHeadIndex]
    infixHeadIndex = find(infixes, infixStart, infixEnd, head)
    
    leftInfixStart = infixStart
    leftInfixEnd = infixHeadIndex - 1

    rightInfixStart = infixHeadIndex + 1
    rightInfixEnd = infixEnd

    print(head, end=' ')

    # postfix head index
    rightPostfixHeadIndex = postfixHeadIndex - 1
    leftPostfixHeadIndex = postfixHeadIndex - (rightInfixEnd - rightInfixStart + 1) - 1

    solve(leftInfixStart, leftInfixEnd, leftPostfixHeadIndex)
    solve(rightInfixStart, rightInfixEnd, rightPostfixHeadIndex)

solve(0, len(infixes) - 1, len(postfixes) - 1)
